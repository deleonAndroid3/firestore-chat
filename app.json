{
  "expo": {
    "name": "Firestore Chat",
    "slug": "firestore-chat",
    "platforms": [
      "ios",
      "android",
      "web"
    ],
    "version": "1.0.0",
    "orientation": "portrait",
    "icon": "./assets/icon.png",
    "splash": {
      "image": "./assets/splash.png",
      "resizeMode": "contain",
      "backgroundColor": "#ffffff"
    },
    "updates": {
      "fallbackToCacheTimeout": 0
    },
    "assetBundlePatterns": [
      "**/*"
    ],
    "ios": {
      "supportsTablet": true
    },
    "description": "POC for Firestore and React Native",
    "githubUrl": "https://github.com/apappas1129/firestore-chat",
    "android": {
      "package": "com.deleonandroid3.firestorechat"
    }
  }
}
