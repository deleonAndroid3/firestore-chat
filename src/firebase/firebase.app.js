/* eslint-disable no-undef */
import firebase from 'firebase'
import '@firebase/auth'
import '@firebase/firestore'
import '@firebase/storage'
import env from '../../environment'

if (!firebase.apps.length) {
  firebase.initializeApp(env.firebaseConfig)
}

const firestore = firebase.firestore()
const cloudStorage = firebase.storage()

/**
 * Creates a new firestore timestamp from the given date.
 * @param {Date} date
 * @returns {Timestamp} A new Timestamp representing the same point in time as the given date.
 * @typedef {import ("firebase").firestore.Timestamp} Timestamp
 */
const getTimestamp = (date) => {
  return firebase.firestore.Timestamp.fromDate(date || new Date())
}

export { firebase, firestore, cloudStorage, getTimestamp }
