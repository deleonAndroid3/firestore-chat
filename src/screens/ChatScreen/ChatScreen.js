import React, { useEffect, useState, useCallback } from 'react'
import { GiftedChat } from 'react-native-gifted-chat'
// import { View, KeyboardAvoidingView, Platform } from 'react-native'
import { View } from 'react-native'
import AsyncStorage from '@react-native-community/async-storage'

import { firestore } from '../../firebase/firebase.app'
import registerToPushNotification from '../../helpers/register-to-push-notification'
import sendPushNotifications from '../../helpers/send-push-notifications'

import PropTypes from 'prop-types'

const chatsRef = firestore.collection('chats')
const expoPushTokensRef = firestore.collection('expoPushTokens')

export function maybeGetEventsByState(state) {  
  let query = firestore.collection('events')
  
  if (state) {
    query = query.where('state', '==', state)
  }
  
  return query.get()
}

const ChatScreen = (props) => {
  const [messages, setMessages] = useState([])
  const [expoPushToken, setExpoPushToken] = useState()
  const [participantTokens, setParticipantTokens] = useState()

  const user = props.userData

  useEffect(() => {
    getExpoNotifToken()

    const unsubscribe = chatsRef.onSnapshot((querySnapshot) => {
      const messagesFirestore = querySnapshot
        .docChanges()
        .filter(({ type }) => type === 'added')
        .map(({ doc }) => {
          const message = doc.data()
          // `createdAt` is firebase.firestore.Timestamp instance
          // https://firebase.google.com/docs/reference/js/firebase.firestore.Timestamp
          return { ...message, createdAt: message.createdAt.toDate() }
        })
        .sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime())

      // Render messages
      appendMessages(messagesFirestore)
    })

    // Stop listening for updates when no longer required
    return () => unsubscribe()
  }, [])

  useEffect(() => {
    const unsubscribe = expoPushTokensRef.onSnapshot((querySnapshot) => {
      const otherTokens = querySnapshot
        .docChanges()
        .filter(({ doc }) => doc.data().token !== expoPushToken)
        .map(({ doc }) => doc.data().token)

      setParticipantTokens(otherTokens)
    })

    return () => unsubscribe()
  }, [])

  const appendMessages = useCallback(
    (messages) => {
      setMessages((previousMessages) =>
        GiftedChat.append(previousMessages, messages)
      )
    },
    [messages]
  )

  async function getExpoNotifToken() {
    let token = await AsyncStorage.getItem('expoPushToken')

    if (!token) {
      registerToPushNotification().then((response) => {
        if (!response) return

        token = response

        // NOTE: We separated the tokens to another collection instead of using the 'users' collection.
        // This is to flatten data and avoid expensive queries when notifying multiple devices.
        expoPushTokensRef.doc(user._id).set({ token })

        AsyncStorage.setItem('expoPushToken', token)
      })
    }

    setExpoPushToken(token)
  }

  async function handleSend(messages) {
    const writes = messages.map((m) => chatsRef.doc(m._id).set(m)) // Can be chatsRef.add(m)) but using it auto generates the document id.

    await Promise.all(writes)

    const notif = {
      title: user.name.trim() + 'sent a message!',
      body: (messages[0].text || '').slice(0, 49), // Android standard max length for notifs is 49 characters
    }
    await sendPushNotifications(notif, participantTokens)
  }

  return (
    // {/* {Platform.OS === 'android' && <KeyboardAvoidingView behavior='padding' />} */}
    <View style={{ flex: 1 }}>
      <GiftedChat messages={messages} user={user} onSend={handleSend} />
    </View>
  )
}

ChatScreen.propTypes = {
  userData:PropTypes.shape({
    _id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired
  }).isRequired,
}

export default ChatScreen