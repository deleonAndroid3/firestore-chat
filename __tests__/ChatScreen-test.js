/* eslint-disable no-undef */

import React from 'react'
import { act } from 'react-test-renderer'
import { shallow } from 'enzyme'
import ChatScreen from '../src/screens/ChatScreen/ChatScreen'
import { firebase } from '../src/firebase/firebase.app'
import FirestoreMock from '../__tests-util__/firestoreMock.mock'


const user = {
  _id: '93nJbIsNNRMezRIXoIgKUg9PKh42',
  email: 'apappas1129@gmail.com',
  name: 'Ts Xs',
}

const chat = shallow(<ChatScreen userData={user} theme={themes[0]}/> )


describe('<ChatScreen />', () => {
  const firestoreMock = new FirestoreMock()
  beforeEach(() => {
    firebase.firestore = firestoreMock
    firestoreMock.reset()
  })

  it('App renders without crashing', async () => {
    await act(async () => {
      expect(chat).toBeTruthy()
    })
  })

  it('App test against snapshot', async () => {
    await act(async () => {
      expect(chat).toMatchSnapshot()
    })
  })

  it('should mock Firestore adding message', (done) => {
    firestoreMock.mockAddReturn = { id: 'test-id' }
    firebase.firestore.collection('chats')
      .add({message: 'Hello World'})
      .then(res => {
        expect(firestoreMock.mockCollection).toBeCalledWith('chats')
        expect(firestoreMock.mockAdd).toBeCalledWith({message: 'Hello World'})
        expect(res.id).toEqual('test-id')
        done()
      })
      .catch(done)
  })

})

