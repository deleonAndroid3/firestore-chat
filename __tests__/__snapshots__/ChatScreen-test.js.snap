// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatScreen /> App test against snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <GiftedChat
    actionSheet={null}
    audioProps={Object {}}
    bottomOffset={0}
    dateFormat="ll"
    disableComposer={false}
    extraData={null}
    forceGetKeyboardHeight={false}
    imageProps={Object {}}
    inverted={true}
    isCustomViewBottom={false}
    isKeyboardInternallyHandled={true}
    isLoadingEarlier={false}
    keyboardShouldPersistTaps="never"
    lightboxProps={Object {}}
    listViewProps={Object {}}
    loadEarlier={false}
    locale={null}
    maxComposerHeight={200}
    maxInputLength={null}
    messageIdGenerator={[Function]}
    messages={Array []}
    minComposerHeight={33}
    minInputToolbarHeight={44}
    onInputTextChanged={null}
    onLoadEarlier={[Function]}
    onLongPress={null}
    onLongPressAvatar={null}
    onPressActionButton={null}
    onPressAvatar={null}
    onSend={[Function]}
    placeholder="Type a message..."
    renderAccessory={null}
    renderActions={[Function]}
    renderAvatarOnTop={false}
    renderBubble={[Function]}
    renderChatEmpty={null}
    renderChatFooter={null}
    renderComposer={null}
    renderCustomView={null}
    renderDay={null}
    renderFooter={null}
    renderInputToolbar={null}
    renderLoadEarlier={null}
    renderLoading={null}
    renderMessage={null}
    renderMessageImage={null}
    renderMessageText={null}
    renderSend={null}
    renderSystemMessage={null}
    renderTime={null}
    renderUsernameOnMessage={true}
    showUserAvatar={false}
    textInputProps={Object {}}
    timeFormat="LT"
    user={
      Object {
        "_id": "93nJbIsNNRMezRIXoIgKUg9PKh42",
        "email": "apappas1129@gmail.com",
        "name": "Ts Xs",
      }
    }
    videoProps={Object {}}
    wrapInSafeArea={true}
  />
</View>
`;
